### DISCUSSION TITLE
Auto off

### DISCUSSION THREAD

**Discussion by drgkt** [arendst/Tasmota#13314] [Q&A] [answered]
Is it possible to program Auto Off 10 minutes after current draw drops to 0.066  amps?

Thanks

**Comment by barbudor** [comment, marked as answer]
Yes with rules
There are different ways but is your teleperiod is the default 300 (5 min) an idea could be to use a trigger 'on tele-energy#current<0.066' can be used to increment 'var1' and 'on var1#state>=2' you can turn power off. 
If you want exactly 10 minutes, then you would need to use RulesTimer

**Reply by sfromis** [reply]
If you use the method with counting while energy consumption is low, remember to reset the counter when consumption level is higher again, and maybe use one-shot or `on var1#state==2` to avoid it triggering repeatedly, possibly including just after again using more energy.

**Comment by drgkt** [comment]
Thank you for the replies but they went over my head since I am a beginner. The device I have is a preflashed one. I control it via web interface (locally) only. I noticed the current value on the web interface when the appliance fed is idle (laser printer).
I need to be guided through step by step if you are so kind, please!

**Reply by barbudor** [reply]
Here are my rules for switching off automatically the power to a PC as soon as the PC has shutdown:
Commands to be adapted for you (and to be entered 1 by 1 in the console)
```
mem1 13                             <- define the power (Watts) level to start the countdown
mem2 16                             <- define the power (W) level to stop the countdown (must be > to Mem1)
mem3 60                             <- define the countdown duration in econds

Rule1 on Energy#Power<%Mem1% do ruletimer1 %Mem3% endon on Energy#Power>%Mem2% do ruletimer1 0 endon
Rule1 1
Rule1 5

Rule2 on rules#timer=1 do power off endon
Rule2 1
```

**Reply by drgkt** [reply]
Thank you **barbudor**

Will this fly? (60 seconds for trial)
Rule1 on Energy#Current[1]<0.067 do ruletimer1 60 endon on rules#timer=1 do power off endon
Rule1 1
Rule1 5

Also the setting "Rule" in this picture, what does it do?
![TimerTasmota](https://user-images.githubusercontent.com/32744663/136920629-adae0be9-c01c-49dd-ae12-803adb03288f.jpg)

**Reply by sfromis** [reply]
You do not need the `[1]` part unless you are using an energy monitoring device with more than one channel. You may want an extra rule to cancel the ruletimer again, if the current climbs above the threshold before the ruletimer expires.

The "Rule" selection for a timer enables you to have a rule trigger fired when the time is reached, like `on Clock#Timer==5 do ... endon` for timer number 5.

**Comment by drgkt** [comment]
Hmmm. The way I have it (I will eliminate [1]), as soon as draw is less than 0.067 amps, the countdown will start and it will NOT stop if in the mean time I send another print job, resulting even in a paper jam, right?

**Reply by sfromis** [reply]
Right, once the current reading dipped low, the short rules example will always cut off after the RuleTimer period no matter what happens. Which may be a mess depending on your hardware, but if moving physical things, it is probably best avoided. Hence my suggestion about resetting the ruletimer if/when the current draw rises again.

**Reply by barbudor** [reply]
@drgkt That's why I have a rule to stop the countdown if power is higher than mem2

It's just my personnal experience with my PCs that having 2 different thredhold is more reliable
May be different for you and a single threshold for both tests may provide satisfying result.

**Reply by barbudor** [reply]
Note that I separated the `on rules#timer==1 do ` into another rule-set without the ONCE mode. I don't remember why but it may totally work also in single rule-set like you did.

**Comment by drgkt** [comment]
Sorry for the delay.
Thank you both.
You have been great help.

I will give it a try with a single rule:
Rule1 on Energy#Current<0.067 do ruletimer1 60 endon on Energy#Current>0.070 do ruletimer1 0 endon on rules#timer=1 do power off endon
Rule1 1
Rule1 5

I was wondering if timer stops running when device is switched off. Because zero current satisfies <0.067, I do not want to find myself in a situation where I issue an ON command and the device shuts off unexpectedly because the timer was running while the device was off.

**Reply by barbudor** [reply]
It is the timer ending that will shutoff the device, so it will not run anymore even if you power again the plug 1 second after it turned off.

**Reply by drgkt** [reply]
Yeah... Suppose timer is 10 min. I manually turn off. Condition <0.067 is met, timer starts. If I manually turn on before 10 min, current will exceed 0.070 so the timer stops.

**Reply by barbudor** [reply]
In case manual power, you could add a rule to reset the timer
`on Power1#state=0 do ruletimer1 0 endon`

**Reply by sfromis** [reply]
You could be unlucky if the energy consumption rises just before the off-interval expired, as there is some lag from consumption level rising till a measurement can have the trigger fire. This is in case of the device consuming the energy going on/off, without this being controlled the relay in the plug.

**Comment by drgkt** [comment]
It worked the 1st time but not after I manually turned it on again.
It should have power off at 22:54 but it did not.
Here is the log.

22:42:26.325 RSL: POWER = ON
22:42:28.671 RUL: ENERGY#CURRENT>0.070 performs "ruletimer1 0"
22:42:28.678 RSL: RESULT = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
22:44:07.069 RUL: ENERGY#CURRENT<0.067 performs "ruletimer1 600"
22:44:07.075 RSL: RESULT = {"T1":600,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}

23:03:52.832 CMD: ruletimer1
23:03:52.839 RSL: RESULT = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}

23:05:24.479 CMD: Rule1
23:05:24.486 RSL: RESULT = {"Rule1":{"State":"ON","Once":"ON","StopOnError":"OFF","Length":172,"Free":339,"Rules":"on Energy#Current<0.067 do ruletimer1 600 endon on Energy#Current>0.070 do ruletimer1 0 endon on Energy#Current==0 do ruletimer1 0 endon on rules#timer=1 do power off endon"}}

**Reply by barbudor** [reply]
In my case, the `on rules#timer=1 do power off endon` is in Rule2 *without* ONCE. Can you check if makes a difference ?

**Reply by drgkt** [reply]
I will. How you undo rule1 5? With rule1 6?

**Reply by barbudor** [reply]
https://tasmota.github.io/docs/Commands/#rule

or `Rule1 4`

**Reply by drgkt** [reply]
Did it your way with 2 rules. Seems to be working so far...
I do not understand why it was not reliable all in one rule...
I will keep you posted.

THANKS AGAIN

**Reply by barbudor** [reply]
The trigger of a ONCE rule needs to be invalid first for it to be triggered again.
Like `on temp>35 do...`, you need to have temp going back below 35° in order to be able to trigger it next time it goes above 35°
With a trigger like `on Rule#Timer=1 do...` it looks like there is no t such invalid state which re-arm the trigger. I ended up like this but didn't bothered looking into the code why. May be some day ...

**Comment by drgkt** [comment]
Although rule1 is set to "once" something seems to be restarting the timer.
Open to suggestions.


09:28:09.198 RSL: POWER = ON
09:28:09.542 RUL: ENERGY#CURRENT>0.070 performs "ruletimer1 0"
09:28:09.549 RSL: RESULT = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
09:29:49.065 RUL: ENERGY#CURRENT<0.067 performs "ruletimer1 600"
09:29:49.072 RSL: RESULT = {"T1":600,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
09:29:54.187 RUL: ENERGY#CURRENT<0.067 performs "ruletimer1 600"
09:29:54.193 RSL: RESULT = {"T1":600,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
09:30:03.684 RUL: ENERGY#CURRENT<0.067 performs "ruletimer1 600"
09:30:03.691 RSL: RESULT = {"T1":600,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
09:30:22.668 RUL: ENERGY#CURRENT<0.067 performs "ruletimer1 600"
09:30:22.674 RSL: RESULT = {"T1":600,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
09:30:27.437 RUL: ENERGY#CURRENT<0.067 performs "ruletimer1 600"
09:30:27.444 RSL: RESULT = {"T1":600,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
09:30:34.686 RUL: ENERGY#CURRENT<0.067 performs "ruletimer1 600"
09:30:34.692 RSL: RESULT = {"T1":600,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
09:30:39.317 RSL: STATE = {"Time":"2021-10-15T09:30:39","Uptime":"39T23:32:47","UptimeSec":3454367,"Heap":26,"SleepMode":"Dynamic","Sleep":0,"LoadAvg":999,"MqttCount":0,"POWER":"ON","Wifi":{"AP":1,"SSId":"SSID","BSSId":"BSSID","Channel":xx,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":2,"Downtime":"0T00:03:16"}}
09:30:39.325 RSL: SENSOR = {"Time":"2021-10-15T09:30:39","ENERGY":{"TotalStartTime":"2021-06-18T21:52:31","Total":0.989,"Yesterday":0.200,"Today":0.011,"Period":1,"Power":9,"ApparentPower":15,"ReactivePower":12,"Factor":0.57,"Voltage":223,"Current":0.067}}


09:43:23.121 CMD: rule1
09:43:23.128 RSL: RESULT = {"Rule1":{"State":"ON","Once":"ON","StopOnError":"OFF","Length":93,"Free":418,"Rules":"on Energy#Current<0.067 do ruletimer1 600 endon on Energy#Current>0.070 do ruletimer1 0 endon"}}

09:56:41.445 CMD: rule2
09:56:41.452 RSL: RESULT = {"Rule2":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":35,"Free":476,"Rules":"on rules#timer=1 do power off endon"}}

**Reply by drgkt** [reply]
I will try to raise the 0.067

**Reply by barbudor** [reply]
it looks lke the current is fluctuating a lot around 0.067 without going above 0.070
Do you need to have so low current ?
I choosed to base my rules on Power because I feel the Power values are easier to manage .

**Reply by drgkt** [reply]
Just changed it, see what happens.
Rule1 on Energy#Current<=0.070 do ruletimer1 600 endon on Energy#Current>0.070 do ruletimer1 0 endon

**Reply by barbudor** [reply]
You can add some "debugging" to see what effective values occurs:
`Rule1 on Energy#Current<=0.070 do backlog ruletimer1 600 ; var9 %value% endon on Energy#Current>0.070 do backlog ruletimer1 0 ; var9 %value% endon`

As %value% will be stored into var9, it will add a log that shows the current value that triggered the rule

**Reply by drgkt** [reply]
Worked as expected without timer restart.
I will report if further errors.

Would not be possible without your help.
Thanks

**Comment by tbone2k-git** [comment]
This is an auto switch off with a delay of 20 seconds, if power consumption was above 16 watts and falls below 8 watts. If power consumption goes back up before the 20s timer expires, auto switch off will be cancelled. Toggling power manually at any time via button or web ui, will also cancel the auto switch off, so it should not switch (off) unexpectedly at any time. Please adjust values 8, 16, 20 in the rule #1 as needed.

```
Rule1 ON Energy#Power>16 DO BACKLOG VAR1 1; RuleTimer1 0 ENDON ON Energy#Power<8 DO IF (Power1#State==on) RuleTimer1 20; ENDIF ENDON
Rule1 1
Rule1 5

Rule2 ON Power1#State DO BACKLOG VAR1 0; RuleTimer1 0 ENDON 
Rule2 1
Rule2 4

Rule3 ON rules#timer=1 DO IF (%VAR1%=1) VAR1 0; Power1 0; ENDIF ENDON
Rule3 1
Rule3 4
````

I post this here, because it took me 3 days to get Tasmota compile and come up with a working set of rules, which don't have weird side effects or don't work at all. I really started to grow grey hair with this. Maybe the next guy can make use of it. You need to compile Tasmota with  #define USE_EXPRESSION and #define SUPPORT_IF_STATEMENT to make these rules work.

All the best and thank you! o)

ps: I use these rules to completely switch off two computer displays, which use 15 watts standby power each. Powering up the monitors is done via http commands to the tasmota devices, when computer wakes up or manually via hotkey on the keyboard.

**Reply by barbudor** [reply]
Good to know that this is doing what you want considering that you are inventing syntaxes that are not supported:
`IF (Power1#State==on)`

**Comment by tbone2k-git** [comment]
Thank you for looking at my post! o) What is the official syntax to check Power1#State in an IF condition?
I tried all the variants like ==1, =1, =on, but had the impression they don't work as expected, or the Power1#State was not updated correctly, since Tasmota was running the following command even though the device was switched off. In hindsight I realized that a lot of issues come with rules in ONCE mode and you have to run "RuleX 1" after updating the rule, else the rule won't run correctly. Maybe I got lost while trying around and mixed things up. If you tell me the correct syntax, I will investigate again. Thank you! o)

**Reply by barbudor** [reply]
There is no syntax for that, you should put the value of Power into a var with 
`ON Power#state do ver9 %value% ENDON`
and test this var `IF (%var9%==0)`

Note that I have 2 rules in order to avoid taking into account 'Power Toggle' (value 2)

**Comment by tbone2k-git** [comment]
"There is no syntax for that.." Really, this makes may day! o) I scrolled the commands/rules references up and down and looked all over the internet to find proper examples on how to use IF and Power1#State, I think I found one at least and it has fooled me? I write code for 25+ years, creating some lines of Tasmota rules seems the hardest thing ever! o)

But fine, I'm willing to learn. o) So I changed the rules to look like that and things seem to work as before.
```
//VAR1 = power consumption is/was above threshold
//VAR2 = helper var to test power state in IF (since Power1#State cannot be used in IF.. directly) 
Rule1 ON Energy#Power>16 DO BACKLOG VAR1 1; RuleTimer1 0 ENDON ON Energy#Power<8 DO IF (%VAR2%=1) RuleTimer1 20; ENDIF ENDON
Rule1 1
Rule1 5

Rule2 ON Power1#State DO BACKLOG VAR1 0; VAR2 %VALUE%; RuleTimer1 0 ENDON 
Rule2 1
Rule2 4

Rule3 ON rules#timer=1 DO IF (%VAR1%=1) VAR1 0; Power1 0; ENDIF ENDON
Rule3 1
Rule3 4

```

> Note that I have 2 rules in order to avoid taking into account 'Power Toggle' (value 2)

I don't understand. You have 2 rules instead of 3 because you don't want a manual power toggle to stop the auto-off timer?

**Reply by barbudor** [reply]
🤣
Tasmota rules are not a programming language in that it doesn't describe a sequence of operations.
It also shaked my brain for quite some time before being use to it.
And still there are things where I still prefer to get an ESP32 and use Berry language because that so much more closer to what I'm used to.

Forget my other sentence. I started to write 2 rules but in fact only 1 is needed because `ON Power1#state DO` react to the result of the power command, not to the incoming Power command itself.

Still what you are writing seems overcomplicated
My current solution is:
```
Rule1 
  on Energy#Power<%Mem1% do ruletimer1 %Mem3% endon 
  on Energy#Power>%Mem2% do ruletimer1 0 endon
Rule1 1
Rule1 5
Rule2 
  on rules#timer=1 do power off endon
Rule2 1
```
I'm using `mem`s to store parameters so I can change them easily without changing the rules
`mem1` is the power to start the countdown to off. In your case 8
`mem2` is the power to stop countdown to on. In your case 16
`mem3` is the timeout
No need for `var` to store the power state.

Edited to fix the typo pointed below

**Reply by marcohald** [reply]
@barbudor Thank you for your example.
I think the last line should be Rule2 1 instead of Rule1 1

**Reply by barbudor** [reply]
Of course
Well spotted
